gapic/samplegen/samplegen.py

```diff
@@ -0,0 +1,376 @@
+        "loop": validate_loop_,
+        "comment": validate_comment_
+    }
+
+
+TransformedRequest = namedtuple("TransformedRequest", ["base", "body"])
+
+
+# TODO: this will eventually need the method name and the proto file
+# so that it can do the correct value transformation for enums.
+def validate_and_transform_request(request: List[Mapping[str, str]]) -> List[TransformedRequest]:
+    """Checks the "request" block from a sample config and returns a transformed version.
+
+       In the initial request, each dict has a "field" key that maps to a dotted
+       variable name, e.g. clam.shell.
+       Multiple dicts may share a variable base name; these indicate that multiple
+       attributes of a request parameter are being configured.
+
+       The only required keys in each dict are "field" and value".
+       Optional keys are "input_parameter" and "value_is_file".
+       All values in the initial request are strings
+       except for the value for "value_is_file", which is a bool.
+
+       The topmost dict of the return value has two keys: "base" and "body",
+       where "base" maps to a variable name, and "body" maps to a list of variable
+       assignment definitions. The only difference in the bottommost dicts
+       are that "field" maps only to the second part of a dotted variable name.
+       Other key/value combinations in the dict are unmodified for the time being.
+
+       E.g. [{"field": "clam.shell", "value": "10 kg", "input_parameter": "shell"},
+             {"field": "clam.pearls", "value": "3"},
+             {"field": "squid.mantle", "value": "100 kg"}]
+              ->
+            [TransformedRequest("clam",
+              [{"field": "shell", "value": "10 kg", "input_parameter": "shell"},
+               {"field": "pearls", "value": "3"}]),
+             TransformedRequest("squid", [{"field": "mantle", "value": "100 kg"}])]
+
+       The transformation makes it easier to set up request parameters in jinja
+       because it doesn't have to engage in prefix detection, validation,
+       or aggregation logic.
+
+
+    Args:
+        request (list[dict{str:str}]): The request body from the sample config
+
+    Returns:
+        list[dict{str:(str|list[dict{str:str}])}]: The transformed request block.
+
+    Raises:
+        DuplicateInputParameter: If an "input_parameter" value is repeated.
+        ReservedVariableName: If an "input_parameter" value or a "base" value
+                              is a reserved word.
+        InvalidRequestSetup: If a dict in the request lacks a "field" key
+                             or the corresponding value is malformed.
+    """
+    base_param_to_attrs: Mapping[str,
+                                 List[Mapping[str, str]]] = defaultdict(list)
+    input_params: Set[str] = set()
+    for field_assignment in request:
+        field_assignment_copy = dict(field_assignment)
+        input_param = field_assignment_copy.get("input_parameter")
```

Comment:
are we also checking for collision between the input param and the locally defined variables in the response? Both sets are in the sample function's scope.

Reply:
Good catch! I don't think the testing for that is sufficient, and so it probably isn't happening.

Reply:
Nor is there an attempt to prevent a collision between the input_params and anything else.
The general lesson seems to be to never, _ever_ accept user input. It's just too much of a hassle :stuck_out_tongue: .

Reply:
Turned request validation/transformation into a method of the validator so that the input variables are added to the variable definitions.